### DISCUSSION TITLE
How to convert PDF to image (jpg, png...) with portable version

### DISCUSSION THREAD

**Discussion by Akxe** [ImageMagick/ImageMagick#2691] [Help] [answered]
Hi, I have a simple question. I need to be able to convert (one page) PDF to image without installing anything to PC running the script.

The restriction is, that we have many (10+) PCs and we would like to avoid software for everyone is bothersome. Also, this would add additional steps to the PC reinstallation process. Also, the script is not on the PC running it, but on network storage.

**Comment by fmw42** [comment]
I do not understand your question. Do you already have Imagemagick? If not, you have to install Imagemagick and Ghostscript delegate. You could install the "magick" portable version.  Otherwise, you will need to find some on-line conversion web site.

**Reply by Akxe** [reply]
I do have the portable version of imagemakick, but I did not find portable version of Ghostscript (that is necessary for pdf)

**Reply by fmw42** [reply]
What do you get from

`magick -version`

Does it list gs or gslib?

What do you get from

`gs --version`

**Reply by Akxe** [reply]
I have magick version of: `ImageMagick 7.0.10-33 Q16 x64 2020-10-04`

But I don't have GhostScript, I don't know how to have it portable

**Reply by fmw42** [reply]
Please show the full output from 

`magick -version`

**Reply by Akxe** [reply]
```
C:\Users\akxe\Downloads\ImageMagick> ./magick.exe -version
Version: ImageMagick 7.0.10-33 Q16 x64 2020-10-04 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 192729111
Features: Cipher DPC HDRI OpenCL
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib
```

**Reply by Akxe** [reply]
> ```
> Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
> ```

Isn't the copyright outdated?

**Comment by fmw42** [comment]
Your portable version of Imagemagick includes gslib in the delegates. That is Ghostscript.  But perhaps you policy.xml file needs to be edited to give read|write permission to use Ghostscript.  See https://imagemagick.org/script/resources.php

**Reply by Akxe** [reply]
That's amazing!

I am getting this error
```
\\SERVER8\Zakazky\ImageMagick\magick.exe '\\soubory8\dokumenty\1-AgE\1 - KLIENTI\starostové\2020\9-2020\Fröhlichová Markéta (510x240) Výlepy.pdf' -resize 150x200 '\\soubory8\dokumenty\1-AgE\1 - KLIENTI\starostové\2020\9-2020\converted.jpg'
magick.exe: FailedToExecuteCommand `"gswin64c.exe" -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 "-sDEVICE=pamcmyk32" -dTextAlphaBits=4 -dGraphicsAlphaBits=4 "-r72x72"  "-sOutputFile=C:/Users/akxe/AppData/Local/Temp/magick-JG_gEClG7yrQT2vwRMF76FYX9IE9XNgx%d" "-fC:/Users/akxe/AppData/Local/Temp/magick-ekmX-qx2r4G3dS67znbY1_NWOWajvzqD" "-fC:/Users/akxe/AppData/Local/Temp/magick-GRKQRJi7mQJlX94wRf1sUedOovpRwzK4"' (Systém nemůže nalézt uvedený soubor.) @ error/delegate.c/ExternalDelegateCommand/475.
magick.exe: PDFDelegateFailed `Systém nemůže nalézt uvedený soubor.' @ error/pdf.c/ReadPDFImage/663.
```

**Reply by t246246** [reply]
If you are sure you have gswin64c.exe somewhere in your server, replace all the instances of "@PSDelegate" in delegates.xml with the full path. See https://stackoverflow.com/questions/13304832/ghostscripts-file-path-in-imagemagick

Also it seems ghostscript dislike backslash: (see https://www.ghostscript.com/doc/current/Use.htm and search "forward slash"), so that might be a trouble.

**Comment by t246246** [comment, marked as answer]
This link might be your help: https://portableapps.com/apps/utilities/ghostscript_portable
I have no experience with this site and their exe files, though, so check it by yourself at your own risks.

**Reply by Akxe** [reply]
> Your portable version of Imagemagick includes gslib in the delegates. That is Ghostscript. But perhaps you policy.xml file needs to be edited to give read|write permission to use Ghostscript. See https://imagemagick.org/script/resources.php

It seems that portable version is bundled with Ghostscript too...

**Reply by t246246** [reply]
I download https://imagemagick.org/download/binaries/ImageMagick-7.0.10-34-portable-Q16-HDRI-x64.zip and checked.
There is no gswin64c.exe in it.

```
$ unzip -l ImageMagick-7.0.10-34-portable-Q16-HDRI-x64.zip
Archive:  ImageMagick-7.0.10-34-portable-Q16-HDRI-x64.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
   104506  2020-10-08 13:43   ChangeLog
     1411  2020-10-08 13:43   colors.xml
 21185744  2020-10-08 13:43   compare.exe
 21185744  2020-10-08 13:43   composite.exe
      757  2020-10-08 13:43   configure.xml
 21185744  2020-10-08 13:43   conjure.exe
 21185744  2020-10-08 13:43   convert.exe
  1040080  2020-10-08 13:43   dcraw.exe
    10038  2020-10-08 13:43   delegates.xml
    53734  2020-10-08 13:43   english.xml
 66603216  2020-10-08 13:43   ffmpeg.exe
 21185744  2020-10-08 13:43   identify.exe
     7637  2020-10-08 13:43   ImageMagick.rdf
 22639824  2020-10-08 13:43   IMDisplay.exe
    12546  2020-10-08 13:43   LICENSE.txt
     2509  2020-10-08 13:43   locale.xml
     1722  2020-10-08 13:43   log.xml
 21185744  2020-10-08 13:43   magick.exe
   135361  2020-10-08 13:43   mime.xml
 21185744  2020-10-08 13:43   mogrify.exe
 21185744  2020-10-08 13:43   montage.exe
    66674  2020-10-08 13:43   NOTICE.txt
     3234  2020-10-08 13:43   policy.xml
     2437  2020-10-08 13:43   quantization-table.xml
     7626  2020-10-08 13:43   README.txt
    60960  2020-10-08 13:43   sRGB.icc
 21185744  2020-10-08 13:43   stream.exe
    11636  2020-10-08 13:43   thresholds.xml
     9470  2020-10-08 13:43   type-ghostscript.xml
      691  2020-10-08 13:43   type.xml
---------                     -------
281447765                     30 files
```

**Reply by Akxe** [reply]
You're right, `gswin64c.exe` is not present...

@fmw42 Is this intended? Should I/You report this as a bug?

**Reply by Akxe** [reply]
> If you are sure you have gswin64c.exe somewhere in your server, replace all the instances of "@PSDelegate" in delegates.xml with the full path. See https://stackoverflow.com/questions/13304832/ghostscripts-file-path-in-imagemagick 

Did install it from that url changed the `delegates.xml` file and it works like a charm. Thanks!

**Reply by Akxe** [reply]
I renamed all (10) occurrences of `@PSDelegate@` in `delegates.xml` in order to make it work. In my example the `command` was like this: `command="&quot;\\\\SERVER8\\Zakazky\ImageMagick\Ghostscript\bin\gswin64c.exe&quot; ..."`

**Comment by snibgo** [comment]
On Windows, ImageMagick installations do not include Ghostscript. Those are available from https://ghostscript.com/download/gsdnld.html

**Comment by Akxe** [comment]
> > If you are sure you have gswin64c.exe somewhere in your server, replace all the instances of "@PSDelegate" in delegates.xml with the full path. See https://stackoverflow.com/questions/13304832/ghostscripts-file-path-in-imagemagick
> 
> Did install it from that url changed the `delegates.xml` file and it works like a charm. Thanks!
